src/main/java/io/github/zap/commons/keyvaluegetters/AbstractKeyValueGetter.java

```diff
@@ -0,0 +1,146 @@
+package io.github.zap.commons.keyvaluegetters;
+
+import io.github.zap.commons.utils.NonParameterizedType;
+import io.github.zap.commons.utils.ReflectionUtils;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Base class for some {@link KeyValueGetter}, contains the default process to locate, deserialize and bind to fields
+ */
+public abstract class AbstractKeyValueGetter implements KeyValueGetter {
+    protected final Builder builder;
+    private final Map<Class<?>, Set<KeyField>> cachedKeyFields = new HashMap<>();
+
+    protected AbstractKeyValueGetter(Builder builder) {
+        this.builder = builder;
+    }
+
+    @Override
+    public <T> T get(Class<T> type) throws GetFailedException {
```

Comment:
i think `type` should be annotated with @NotNull and this goes for any publicly accessible parameter that will throw a NullPointerException/result in unexpected behavior if null is passed